packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts

```diff
@@ -239,4 +242,10 @@ export const useGridSelection = (apiRef: GridApiRef): void => {
     });
     forceUpdate();
   }, [apiRef, setGridState, forceUpdate, isRowSelectable]);
+
+  React.useEffect(() => {
+    if (totalSelectedRows > Object.keys(visibleRowIds).length) {
```

Comment:
you can use `visibleGridRowCountSelector`
Also not sure about the logic here. What if we change rows to another set with the same length?

Reply:
> Also not sure about the logic here. What if we change rows to another set with the same length?

So when filtering, is it possible that the rows filtered (visible) are equal to the total rows? Also this logic only runs when the rows are _selected_ and then the filtering is applied, so the visible rows are less after filtering. Would there be a case when the rows are of same length after filtering? I don't think so.

Reply:
> you can use visibleGridRowCountSelector

I need to pass `visibleRowIds` below to `selectRows`. Should I still select `visibleRowCount` for count to replace `Object.keys(visibleRowIds).length`? I think we should as the selector will be memoized.

Reply:
> So when filtering, is it possible that the rows filtered (visible) are equal to the total rows?

Yes, if all rows satisfy the filter than all rows will be visible.

---

You can't just check if there're more selected rows than visible rows. The behavior becomes inconsistent if the size of the selection is smaller than the number of visible rows.

I think you need to do the same thing we do when the `rows` prop is changed, but with `visibleRowIds`.

https://github.com/mui-org/material-ui-x/blob/a2a1f8eca0f8b188a1ed40abbcc616dc3a837d76/packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts#L203-L208

@DanailH was working in https://github.com/mui-org/material-ui-x/pull/1864 to fix a bug around this part of the code too.

Reply:
> So when filtering, is it possible that the rows filtered (visible) are equal to the total rows?
> > Yes, if all rows satisfy the filter then all rows will be visible.

Okay, so then the point is that this condition won't satisfy, right? And it would not select rows by deselecting others and selecting visible ones. It would work as it works now in master.

> The behavior becomes inconsistent if the size of the selection is smaller than the number of visible rows.

Again, then this effect's condition won't satisfy and selecting rows won't run.

------

**What this effect does?**

I will like to give an example: 
For grid with 200 rows.
The total selected rows are all 200, after filter the visible rows become 40. The dependencies change, this effect's condition satisfies and now the select rows will deselect others and select only 40. If you add another filter, the rows are now 37, it will select only 37. This is main issue of #1141.

-----

> I think you need to do the same thing we do when the `rows` prop is changed, but with `visibleRowIds`.
>
> https://github.com/mui-org/material-ui-x/blob/a2a1f8eca0f8b188a1ed40abbcc616dc3a837d76/packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts#L203-L208
>
> @DanailH was working in https://github.com/mui-org/material-ui-x/pull/1864 to fix a bug around this part of the code too. 

Does this effect #1141? Can we focus on the main issue and the related comments issues I linked in description?

Also I would suggest once to look at whats happening currently and after the fix for the issues attached in the description.

Reply:
> you can use visibleGridRowCountSelector

@dtassone I used this selector now.

Reply:
> Again, then this effect's condition won't satisfy and selecting rows won't run.

Yeah, but you might have one selected row, adds another filter where this selected row doesn't satisfy it and the previous row will still selected. If you select all, that means having a larger selection than visible rows, it unselects those rows that are not more visible. That's the inconsistency I noticed.

> Does this effect #1141? 

No

> Is this fix you are suggesting for the related issue of #1141?

Yes, I would keep it simple. The selection should always contain only visible rows. In the future, we can add an option to allow to accumulate selection even when rows become invisible.

Reply:
> Yeah, but you might have one selected row, adds another filter where this selected row doesn't satisfy it and the previous row will still be selected.

Yeah, even I see this behaviour. 

> The selection should always contain only visible rows

Are you referring with pagination, that it should select only the current page visible rows? #605?